### PR TITLE
nixos/haproxy: allow to specify haproxy package

### PR DESCRIPTION
**Pull request by jvanbruegge**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jvanbruegge** [comment]
The module also uses the "haproxy" user currently, it would be better to use Systemd's DynamicUser for this. Should I change this as well?

File: nixos/modules/services/networking/haproxy.nix [hunk 1]
```diff
@@ -26,6 +26,13 @@ with lib;
         '';
       };
 
+      package = mkOption {
+        type = types.package;
+        default = pkgs.haproxy;
+        defaultText = literalExpression "pkgs.haproxy";
+        description = lib.mdDoc "HAProxy package to use.";
+      };
+
```
Comment (drupol) [inline nixos/modules/services/networking/haproxy.nix:35]: Use `mkPackageOptionMD` like in https://github.com/NixOS/nixpkgs/pull/237824/files#diff-54db11ee9dbd51466b032f18228d1ddc8da76ada6fb2306ccffb7e7b8cfc4f8dR14

File: nixos/modules/services/networking/haproxy.nix [hunk 2]
```diff
@@ -70,15 +77,15 @@ with lib;
...
+          "${pkgs.coreutils}/bin/ln -sf ${cfg.package}/sbin/haproxy /run/haproxy/haproxy"
           # when running the config test, don't be quiet so we can see what goes wrong
           "/run/haproxy/haproxy -c -f ${haproxyCfg}"
         ];
         ExecStart = "/run/haproxy/haproxy -Ws -f /etc/haproxy.cfg -p /run/haproxy/haproxy.pid";
         # support reloading
         ExecReload = [
-          "${pkgs.haproxy}/sbin/haproxy -c -f ${haproxyCfg}"
-          "${pkgs.coreutils}/bin/ln -sf ${pkgs.haproxy}/sbin/haproxy /run/haproxy/haproxy"
+          "${cfg.package}/sbin/haproxy -c -f ${haproxyCfg}"
```
Comment (drupol) [inline nixos/modules/services/networking/haproxy.nix:87]:
```suggestion
          "${lib.getExe cfg.package} -c -f ${haproxyCfg}"
```

File: nixos/modules/services/networking/haproxy.nix [hunk 3]
```diff
@@ -70,15 +77,15 @@ with lib;
...
+          "${pkgs.coreutils}/bin/ln -sf ${cfg.package}/sbin/haproxy /run/haproxy/haproxy"
```
Comment (drupol) [inline nixos/modules/services/networking/haproxy.nix:88]:
```suggestion
          "${pkgs.coreutils}/bin/ln -sf ${lib.getExe cfg.package} /run/haproxy/haproxy"
```

File: nixos/modules/services/networking/haproxy.nix [hunk 4]
```diff
@@ -70,15 +77,15 @@ with lib;
         ExecStartPre = [
           # when the master process receives USR2, it reloads itself using exec(argv[0]),
           # so we create a symlink there and update it before reloading
-          "${pkgs.coreutils}/bin/ln -sf ${pkgs.haproxy}/sbin/haproxy /run/haproxy/haproxy"
+          "${pkgs.coreutils}/bin/ln -sf ${cfg.package}/sbin/haproxy /run/haproxy/haproxy"
```
Comment (drupol) [inline nixos/modules/services/networking/haproxy.nix:80]:
```suggestion
          "${pkgs.coreutils}/bin/ln -sf ${lib.getExe cfg.package} /run/haproxy/haproxy"
```

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution, I've left some comments, let me know what you think.

**Comment by jvanbruegge** [comment]
done

**Review by drupol — APPROVED** [review]

**Pull request merged**